File: fsharp/Allpairs.fs
(*
 File: Allpairs.fs
 Description: nbody computation: barnes-hut algoritms
 Author: Prabhat Totoo
*)

module Allpairs

open System
open System.IO
open System.Diagnostics

open MapSkel

// (x,y,z,vx,vy,vz,m)
type Body = Body of double * double * double * double * double * double * double

// (ax,ay,az)
type Accel = Accel of double * double * double

let timeStep = 0.001
let eps = 0.01

let getBodies n generate =
    if generate then
        let newBody seed =
            let rnd = new Random(seed)
            let [x;y;z;vx;vy;vz;m] = List.init 7 (fun _ -> rnd.NextDouble())
            Body (x,y,z,vx,vy,vz,m)
        Seq.map newBody [0..(n-1)]
        |> Seq.toList
    else
        let bodyFromLine (line:string) =
            let words = line.Split [|' '|]
            let word (i:int) = double (Seq.nth i words)
            Body ((word 1),(word 2),(word 3),(word 4),(word 5),(word 6),(word 0))
        File.ReadLines("input.txt")
        |> Seq.map bodyFromLine
        |> Seq.toList

// checking function
let f acc (Body (x,y,z,vx,vy,vz,m)) = acc+x+y+z+vx+vy+vz+m

let mutable v = 0

let rec doSteps s bs =
    match s with
    | 0 -> bs
    | _ ->
        let updateVel (b:Body) =
            let accel (Body (ix,iy,iz,_,_,_,_) as bi) (Body (jx,jy,jz,_,_,_,jm) as bj) =
                if (bi = bj) then Accel (0.0,0.0,0.0)
                else
                    let dx = ix - jx
                    let dy = iy - jy
                    let dz = iz - jz
                    let dSquared = dx*dx + dy*dy + dz*dz + eps
                    let distance = sqrt (dSquared)
                    let mag = timeStep / (dSquared * distance)
                    let mmag = jm*mag
                    Accel ((dx*mmag),(dy*mmag),(dz*mmag))
            let deductChange (Body (x,y,z,vx,vy,vz,m)) (Accel (ax,ay,az)) = Body (x,y,z,(vx-ax),(vy-ay),(vz-az),m)
            
            //List.fold deductChange b (List.map (accel b) bs)  //mmxp230: 25.76s
            //fold/map merge
            List.fold (fun state b' -> deductChange state (accel b b')) b bs  //mmxp230: 14.45s

        let updatePos (Body (x,y,z,vx,vy,vz,m)) = Body ((x+timeStep*vx),(y+timeStep*vy),(z+timeStep* z),vx,vy,vz,m)
        let mapFunc =
            match v with
            | 0 -> List.map
            | 1 -> pmap_async_lst
            | 2 -> pmap_plinq_lst
            | _ -> List.map // default is sequential map 
        let new_bs = mapFunc (updatePos << updateVel) bs
        doSteps (s-1) new_bs

let run n steps pv =
    v <- pv
    printfn "ver: Allpairs \nparver: %d \nn: %d \nsteps: %d" v n steps
    let bs = getBodies n true
    let res1 = List.fold f 0.0 bs
    printfn "%f" res1
    let t1 = Stopwatch.StartNew()
    let bs' = doSteps steps bs
    let res2 = List.fold f 0.0 bs'
    printfn "%f" res2
    let t2 = t1.Elapsed.TotalSeconds
    printfn "time taken: %.2fs" t2

File: scala/allpairs.scala
// Time-stamp: <Sat May 26 03:16:32 BST 2012 pt114>
import scala.actors.Futures
import scala.io.Source
import scala.util.Random
import scala.annotation.tailrec

object allpairs {
	class Body (val mass: Double, val x: Double, val y: Double, val z: Double, val vx: Double, val vy: Double, val vz: Double)
		
	val timeStep = 0.001
	val eps = 0.01

	// Returns nbodies either from generator or input file.
	def getBodies(num: Int, generate: Boolean): List[Body] = {		
		if (generate) {
			def newBody(seed: Int) = {
				val rnd = new Random(seed)
				new Body(rnd.nextDouble(),rnd.nextDouble(), rnd.nextDouble(), rnd.nextDouble(),rnd.nextDouble(), rnd.nextDouble(), rnd.nextDouble())
			}
			
			for (i <- List.range(0, (num-1))) yield (newBody(i))
		}
		else {
			Source.fromFile("input.txt").getLines().toList
				.map((x: String) => x.split(" ").map((y: String) => y.toDouble))
				.map((arr: Array[Double]) => new Body(arr(0),arr(1), arr(2), arr(3),arr(4), arr(5), arr(6)))
		}
	}
	
	def f(bs: List[Body]): Double =
		bs.foldLeft(0.0) ((acc: Double, b: Body) =>
			acc + b.x + b.y + b.z + b.vx + b.vy + b.vz + b.mass)
			
	def main(args: Array[String]) = {
		val n = args(0).toInt
		val steps = args(1).toInt
		val bs = getBodies(n, true)		
		println(f(bs))		
		val startTime = System.currentTimeMillis
		val new_bs = doSteps(steps, bs)
		println(f(new_bs))
		val endTime = System.currentTimeMillis-startTime
		println("\ntime taken: (" + endTime / 1000.0 + " s)")
    }
    
	def doSteps(s: Int, bs: List[Body]): List[Body] = s match {
		case 0 => bs
		case _ => {
		
			def updateVel(b: Body) : Body = {
				/* //45.09s
				bs.map((b2: Body) => accel(b,b2)).foldLeft(b) ((acc,a) =>
					new Body (acc.mass,acc.x,acc.y,acc.z,(acc.vx-a._1),(acc.vy-a._2),(acc.vz-a._3))
				)*/
				//fold/map merge
				bs.foldLeft(b) ((state,b2) => {
					val (ax,ay,az) = accel(b,b2)
					new Body (state.mass,state.x,state.y,state.z,(state.vx-ax),(state.vy-ay),(state.vz-az))
				})				
			}
			
			def accel(bi:Body,bj:Body):(Double,Double,Double) = {
				if (bi == bj) (0.0,0.0,0.0)
				else {
					val dx = bi.x - bj.x
					val dy = bi.y - bj.y
					val dz = bi.z - bj.z
					val dSquared = dx*dx + dy*dy + dz*dz + eps					
					val distance = math.sqrt (dSquared)
					val mag = timeStep / (dSquared *  distance)
					val mmag = bj.mass*mag
					((dx*mmag),(dy*mmag),(dz*mmag))
				}
			}
			
			def updatePos(b: Body) : Body =
				new Body(b.mass,(b.x + timeStep * b.vx) , (b.y + timeStep * b.vy), (b.z + timeStep * b.vz),b.vx,b.vy,b.vz)
			
			val new_bs = bs.map(updateVel _ andThen updatePos)
			doSteps(s - 1, new_bs)
		}
	}
}
